### ISSUE TITLE
Rework `oxc_prettier`

### ISSUE THREAD

**Issue by Boshen** [oxc-project/oxc#5068]
[`crates/oxc_prettier`](https://github.com/oxc-project/oxc/tree/main/crates/oxc_prettier) was my attempt at the [prettier bounty](https://console.algora.io/challenges/prettier). 

I thought I could finish it in time, except the fact that I rushed too quickly without looking at all the requirements ... It was too late when I got blocked by printing comments.

In order to rework `oxc_prettier`, we need to understand at least:

* How to debug prettier
* The [`Doc`](https://github.com/prettier/prettier/tree/main/src/document) IR https://github.com/prettier/prettier/blob/main/commands.md
* How to handle comments https://github.com/prettier/prettier/blob/main/src/language-js/comments/handle-comments.js
* How to handle parentheses https://github.com/prettier/prettier/blob/main/src/language-js/needs-parens.js

As for the infrastructure, we already have most of the code:

* https://github.com/oxc-project/oxc/tree/main/tasks/prettier_conformance shows our js compat is at js compatibility: 264/586 (45.05%). But we need to update prettier to latest make sure this conformance runner still works.
* All the code in https://github.com/oxc-project/oxc/tree/main/crates/oxc_prettier mimics and ports prettier https://github.com/prettier/prettier/tree/main/src/language-js

Feel free to remove everything and start from scratch, and copy over the format code https://github.com/oxc-project/oxc/tree/main/crates/oxc_prettier/src/format

**Comment by parkin-lin** [comment]
I'm sorry to take up some of your time, but I feel confused about the following issues, so please allow me to ask you a few questions...
As we konw, [the next minor release of ESLint will deprecate core formatting rules. ](https://eslint.org/blog/2023/10/deprecating-formatting-rules) But this also means that the linter itself originally had code formatting capabilities, so why isn't Oxlint **directly** compatible with plugins like [ESLint Stylistic](https://github.com/eslint-stylistic)? Are there some corner cases where Oxlint doesn't work as well as Prettier, so Oxlint had to be compatible with ESLint and Prettier **separately**, or it's just a divide-and-conquer engineering trade-off？
I'm guessing one of the reasons for this is that Prettier can support different file formats (not only .js, but also .css and more...), but ESLint or Oxlint can do the same job via plugins.
Thanks.

**Comment by DonIsaac** [comment]
We made an intentional decision to only support code correctness rules, not stylistic rules. @Boshen can explain reasoning more.

**Comment by Boshen** [comment]
@leaysgur is writing a series of articles in preparation of this task:

* [Prettier のコードを読む Part.3](https://leaysgur.github.io/posts/2024/09/04/143544/)
* [Prettier のデバッグ方法について](https://leaysgur.github.io/posts/2024/09/03/111109/)
* [Prettier のコードを読む Part.2](https://leaysgur.github.io/posts/2024/09/02/144240/)
* [Prettier のコードを読む Part.1](https://leaysgur.github.io/posts/2024/09/02/103846/)

**Comment by leaysgur** [comment]
> I'm also working on comment attachments to unblock prettier.

Does this mean that you aim to implement Prettier equivalent which achieves with 60+ utils? 🫨 

FYI: Babel also has relevant code(just 300 loc), but that did not seem to meet Prettier's requirement.

**Comment by Boshen** [comment]
> Does this mean that you aim to implement Prettier equivalent which achieves with 60+ utils? 🫨

We need to figure out what exactly is prettier doing with those 60+ utils 🥲

**Comment by IWANABETHATGUY** [comment]
For those who are interested in algorithms under the hood, prettier is based on  https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf,

**Comment by leaysgur** [comment]
It has been 3 weeks since I started reading the Prettier source code. 
It's still far from being complete, but I'd like to leave some progress and summary here.


> How to debug Prettier

There are 3 ways:

- Playground
  - Enable "Debug" options in the sidebar
- CLI
  - With `--debug-*` args
- Node.js API
  - Under `__debug` exports

> https://leaysgur.github.io/posts/2024/09/03/111109/ 

It is written in Japanese, but it is all code, so you can understand it. 😉 

I also recommend to run `node --inspect-brk` the code with `debugger` and inspect it from Chrome DevTools.

> How to handle comments

I will post some topics for discussion in a few days.

**Comment by leaysgur** [comment]
> How to handle comments

As you may know, Prettier's formatting process consists of roughly 3 phases:

- P1: original text > AST
- P2: AST > Doc
- P3: Doc > formatted text

Comments are collected in P1 and used in P2.

In P1:

- It simply parses to the AST with comments in the output
  - However, there are also some adjustments made to the AST nodes
  - Some parsers, such as Babel, already attach comments to nodes at this stage
    - However, Prettier does not use them
    - The reason is to support parsers other than Babel
- As the final step of P1 (more technically, the beginning of P2), comments are classified and attached to nodes
  - First, it starts by finding nearby nodes for each comment
  - Based on that, it determines the placement(ownLine, endOfLine, remaining) from the lines before and after each comment
  - Then, it handles about 30(!) known special patterns for each placement
  - Finally, it finishes using unique tie-breaking algorithm

As a result, some AST nodes have `comments` property with array of `Comment` extended with `leading`, `trailing` and few more props.

In P2 (I haven’t read the code in detail here yet),

- When each node is converted into a Doc, comments are also converted into Docs
  - Therefore, how they are output seems to have already been decided in P1

In OXC, part of the necessary information is already implemented and can be obtained. / #5785

However, just like with Babel, that information may be different from what Prettier requires...

---

So, I think I’ve generally understood "what" Prettier is doing.

However, as for "why" Prettier does it that way, I can only say it’s because that’s Prettier’s opinion.

Incidentally, there seem to be at least around 120 issues related to JS/TS at the moment, but

> https://github.com/prettier/prettier/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+label%3Alang%3Ajavascript%2Cjsx%2Ctypescript+label%3Atype%3Abug

about 50 of them are related to comments, with some remaining unresolved since as far back as 2017.

> https://github.com/prettier/prettier/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+label%3Alang%3Ajavascript%2Cjsx%2Ctypescript+label%3Atype%3Abug+label%3Aarea%3Acomments

**Comment by leaysgur** [comment]
So, what I would like to discuss(confirm?) is: where should `oxc_prettier` aim to go?
As the name suggests, should it strive for 100% compatibility with Prettier?

If the goal is to achieve compatibility, I think it would mean porting almost all of Prettier’s lines.

- Prettier’s coverage is impressively high at 99%.
  - https://app.codecov.io/gh/prettier/prettier/tree/main/src
- Since we don’t use CST like Biome, many of them cannot be omitted
- We would also end up porting the same issues I mentioned above

I believe the original aim of this issue was this, but again, is this an acceptable?

If we decide to proceed, let’s consider how we might specifically approach it...

Fortunately, the 3 phases have isolated inputs and outputs, so:

- If we have AST + Comments (in .json), we can create a Doc
- If we have a Doc (in .json), we can generate formatted text

I think reducing the gaps between each phase will be important.

For that, we’ll need some way to directly compare the result of `Prettier.parse()` with our implementation.

- For example, we could resolve the issue of making our AST ESTree-compatible first?
  - Fortunately, we know Prettier can run without `line|column`
    - https://gist.github.com/leaysgur/411636c61f20558ced9e08a394dc7584
- Alternatively, complete the experiment of aligning with Babel AST?
  - https://github.com/leaysgur/prettier?tab=readme-ov-file
  - Prettier supports not only ESTree but also Babel AST directly, and OXC AST has many similarities with Babel AST

```rs
// OXC
let oxcASTWithComments = oxc_prettier::parse(text);
let oxcEstreeASTWithComments = estreeJSON(oxcASTWithComments);
// let oxcBabelASTWithComments = babelJSON(oxcASTWithComments);

// Prettier
let prettierEstreeASTWithComments = Prettier.parse(text, { parser: "meriyah" });
// let prettierBabelASTWithComments = Prettier.parse(text, { parser: "babel" });

// Diff!
assert(oxcEstreeASTWithComments, prettierEstreeASTWithComments);
```

What if we don’t aim for full compatibility?

- For example, we could continue to use Prettier just as a reference?
  - But handle comments in our own way without focusing on compatibility
  - In my opinion as a user, it would be fine to simply restore comments as close to their original positions as possible without special handling
    - (More personally, I also dislike when parentheses are removed without permission)

But in that case, maybe it would be more like `oxformat` rather than `oxc_prettier`?

Anyway, this is what I was thinking these days. What do you think?

**Comment by Boshen** [comment]
For the long run, I envision a more configurable and less opinionated formatter.

Under this assumption, we can relax the "100% compatibility" constraint.

My intention is to reach a high compatibility with prettier so we can bootstrap ourself, and then start deviating behavior.

I've also looked at the prettier and rustfmt issues before, comment positioning is a really difficult subject due to the natural of matching the original intent of the comment.

To move things forward, I suggest @leaysgur to start refactoring our prettier code in preparation for what to come, I believe you know what about prettier than I do. You may ask @Sysix for help given they have started working on this area as well.

Our current code for P2 and P3 is also incomplete, I also suggest to do a little bit of rework first, as well as adding more debug utilities.

**Comment by Boshen** [comment]
> I think reducing the gaps between each phase will be important.
> For that, we’ll need some way to directly compare the result of Prettier.parse() with our implementation.

I don't think we need to do this. This will leave us too coupled with the prettier implementation, and it may end up being a waste of effort.

We are already matching half of the generated text, a few more iterations of refactoring and implementing details should close our gap to prettier.

**Comment by leaysgur** [comment]
I see, what I likely concerned about was: what exactly is meant by "high compatibility". 😅 
Does it have to be 100%, or should aim for at least 80% if possible, or is something around 70% good enough? etc...

So, It's good to know that we're not necessarily aiming for 100% compatibility.

It's still uncertain how much the percentage will drop if we give up on porting completeness for comment handling, but for now, I'll move forward.

Following your suggestion:

- Understanding and refactoring the `oxc_prettier`
- Continuing with code reading of Prettier from P2

I’ll work on these whenever I have time! 🚀 

---

> For the long run, I envision a more configurable and less opinionated formatter.

I love this idea!

**Comment by magic-akari** [comment]
> For the long run, I envision a more configurable and less opinionated formatter.

Does this mean that oxc_prettier will provide a wide range of configurable options, and offer a preset called `prettier` to match with prettier?

**Comment by Boshen** [comment]
@leaysgur I was just informed that [`biome_formatter`](https://github.com/biomejs/biome/tree/main/crates/biome_formatter) is a port of the Prettier IR and printer infrastructure. It may be possible to reduce repeated work by using `biome_formatter`. More investigation for the rework!

So instead of using our current unfinished IR and printer, we can replace them with `biome_formatter`.

**Comment by leaysgur** [comment]
Thanks for the pinning!

> biome_formatter is a port of the Prettier IR and printer infrastructure.

Yes, and I just started reading through `biome_formatter` and `biome_js_formatter`, so the timing is perfect. 😄

I hadn’t considered it as a viable option for OXC (though I’m not entirely sure why), but if we shift our goal to generating `biome_formatter`'s IR, it could significantly reduce the amount of work, maybe?

I’ll look into this soon (after finishing the regex parser rework), including:

- General architecture: https://github.com/biomejs/biome/blob/main/crates/biome_formatter/CONTRIBUTING.md
- How their CST(especially comments) is used
- `biome_formatter` IR, aka `format_element`
- etc...

**Comment by Boshen** [comment]
> but if we shift our goal to generating biome_formatter's IR

> biome_formatter IR, aka format_element

I think it's the same thing as `prettier`s IR. Also `Doc` in our code.

**Comment by srijan-paul** [comment]
Hey all, just dropping by!

> My intention is to reach a high compatibility with prettier so we can bootstrap ourself, and then start deviating behavior.

I feel adding support for comments to the current implementation should be feasible.
In fact, I suspect it'll give us a significant boost on the prettier conformance test suite from the current 39%.
We don't have to follow prettier's implementation exactly, but we can do something close to it.

Since the oxc_ast's `Program` node already stores a `pub comments: Vec<'a, Comment>` field,
we could just repurpose it to add a third pass over the `Doc` objects (before the `Command`s are generated)?
I think it'd end up looking similar to what @leaysgur showed [here](https://github.com/oxc-project/oxc/issues/5068#issuecomment-2382171971).

I could do a PoC on my fork and have it up in some time.
Would any of you be interested in taking a look, perhaps?

**Comment by pumano** [comment]
I want to share my overall idea from end user perspective: After oxc_prettier will be created I suggest to create something like `eslint-plugin-oxc-prettier` in future for using `oxlint` with that plugin to show problems in IDE / console and fix it as eslint-plugin-prettier does.

**Comment by YurySolovyov** [comment]
you might want to consider running prettier (or oxc_prettier later) separately:
https://www.joshuakgoldberg.com/blog/you-probably-dont-need-eslint-config-prettier-or-eslint-plugin-prettier/

**Comment by leaysgur** [comment]
Recently, I've been reading and researching the code related to Biome's Formatter.
I'd like to note some of my learnings and progress here.

### `biome_formatter` and `biome_xxx_formatter`

In the Biome repo, the `biome_formatter` crate is used as infrastructure to implement Formatters for each language.

According to [CONTRIBUTING.md](https://github.com/biomejs/biome/blob/main/crates/biome_formatter/CONTRIBUTING.md), the implementation on each language side(e.g. for JS) is as follows:

- Use the `biome_rowan` crate to define AST/CST nodes as `biome_js_syntax`
  - Also used as the parsing result of `biome_js_parser`
- In the `biome_js_formatter` crate:
  - Implement the `FormatRule` trait (compatible with the `format!` macro) for all those nodes
  - Implement `JsFormatLanguage` that implements the `FormatLanguage` trait
    - Contains information like context and handling of comments
- Then, call `biome_formatter::format_node(parsed_tree, format_language)`
  - Performs preprocessing on the AST/CST, finds attached comments, and converts to IR
  - By calling `print()` on the return value, you get the formatted string

If I understand correctly, to fully benefit from `biome_formatter`, we need the AST/CST defined using `biome_rowan`.

In the Biome repo, that's natural, but since we want to use OXC's AST, we cannot choose this way.

### `biome_formatter::Printer`

As another way to utilize the `biome_formatter`, using only its IR and `Printer` may be possible.

A concern is that the IR and its `builders` also depend on `biome_rowan`.
Possibly, some of them may not be usable, and in the worst case, there may be no alternative.

Due to my lack of my Rust experience, I couldn't make this judgment from the code and types alone, I think I need to write some code to verify it. I might find problems in more fundamental parts...🥲

Also in this case, the formatting options are limited to those defined in `PrinterOptions`.

- indent_width
- print_width
- line_ending
- indent_style
- attribute_position
- bracket_spacing

> https://github.com/biomejs/biome/blob/main/crates/biome_formatter/src/printer/printer_options/mod.rs#L8

If we want to perform more flexible formatting, we'll have to fork it.

### Biome's IR

Biome's IR is called `FormatElement`, and kinds are:

- Space
- HardSpace
- Line
- ExpandParent
- StaticText
- DynamicText
- LocatedTokenText
- LineSuffixBoundary
- Interned
- BestFitting
- Tag

> https://github.com/biomejs/biome/blob/main/crates/biome_formatter/src/format_element.rs#L19

They are similar but not the same as Prettier's `Doc`.
Biome also has a struct called `Document`, but this just holds a `Vec<FormatElement>`.

> https://github.com/biomejs/biome/blob/main/crates/biome_formatter/src/format_element/document.rs#L17

Although the final formatting result is compatible with Prettier, it's better to think of the intermediate processes are entirely different(more sophisticated).

### Formatting flow

Basically proceeds in the same way as Prettier.

- Parse the code string into an AST/CST
- Preprocess the parsed tree
  - Normalize `()` and logical expressions, just like Prettier
- Collect comments
  - Checking line spacing between comments and nodes, as in Prettier
  - Determining the `placement` of comments also requires heavy conditional branching
- Traverse the tree and convert it into `Vec<FormatElement>`
  - Each node knows its own formatting rules
  - Comments are converted similarly
- Convert `Document::from(format_elements)` into a code string

A part different from Prettier is the determination of whether `()` are necessary.
In the case of Biome, it seems that information is held on the node definition side of `biome_js_syntax`.

> https://github.com/biomejs/biome/tree/main/crates/biome_js_syntax/src/parentheses

As for comments, it seems a lot of code is needed, with both `biome_js_formatter` and `biome_formatter` having over 1,000 lines of code.

If we declare Prettier compatibility, it means our codebase will also need a corresponding amount of code...

---

As the next step, I'm planning to write some code to see whether using only `Printer` is feasible.

I'm also thinking, "Should we aim for Prettier-compatible comment formatting by creating/maintaining all that code again?" ... but I'm not sure.😌

**Comment by leaysgur** [comment]
Progress: An experiment to see if we can use just the `Printer` from `biome_formatter`

Based on `biome_formatter::format_node()`, here's a minimal code example that reproduces a similar flow.

> https://gist.github.com/leaysgur/de62398266ef98edcf7a0e3c06325538

My personal impression is that it doesn't seem impossible, but I'm not yet sure if it's the right path. 😅

Here's what I've found out:

- To use `Printer`, create it with `Formatted::new(document, context)` and then call `print()`
  - A `Document` can be created from a `Vec<FormatElement>`
- In the original code, `document.propagate_expand()` is called before creating `Formatted`
  - However, this `propagate_expand()` is `pub(crate)`, so it can't be used for us(❗️)
  - It seems it would work if you copy it entirely and adjust the types, but...

- `FormatElement` is constructed using helpers under `builders::*`
- Regarding text printing, there are 3 types of IR and helpers:
  - `StaticText`: Accepts `&'a static` only
  - `DynamicText`: Accepts `&str`, but requires a starting position by `biome_text_size::TextSize`
  - `LocatedTokenText`: Requires a `biome_rowan::SyntaxToken`
- In OXC's AST, tokens can only be generated using `StaticText`
  - Because there's no AST corresponding to `new` in `new Foo()`
- Dynamic parts are often represented with `Atom<'a>`, so they can't be `&'a static`
  - Of course, they can't be `SyntaxToken` either
  - That means we have to use `DynamicText`, but for that, we need to provide a `TextSize`
  - While the `start` of `Span` fits the type, internal behavior and policy differences may affect things(❗️)
    - In the first place, I'm not sure what is the sourcemap used for in the Formatter

I mentioned "minimal" at the beginning because this Gist does not use the `Format` trait, which is center of the `biome_formatter`.

- We cannot directly implement the trait from the `biome_formatter` crate in the `oxc_ast` crate
  - Due to Rust's orphan rule
- To cope with this, it requires a lot of boilerplate code
  - While it's understandable that Biome needs these as common infrastructure, but how about for us?
- To refer to Biome's code and use other helpers, it might be better to implement it
  - There are parts we can avoid even with `format_with()`, but still

Next, I will check this part.

**Comment by Boshen** [comment]
@leaysgur Thank you so much for the research. I think it's enough evidence to just implement our own thing (with some parts ported from Biome)?

**Comment by leaysgur** [comment]
> to just implement our own thing

Yes..., I also think that's probably the case.

And if we choose to go our own way, even if there are parts we can refer to, I feel there won't be any parts where we can borrow as code. 😓

Go back to the square one, I will re-start by understanding/refactoring the current code, and then think about what to do next.

**Comment by Boshen** [comment]
@leaysgur will lead this project, and is free to make any changes.

**Comment by leaysgur** [comment]
Thanks~!

Progress is likely to be slow due to limited resources, but I will try to update progress every 2-3 days or so at the latest comment on this issue.

- - -

Now I'm working on:

- Reading current `oxc_prettier` source
- Refactoring macro usage, `Doc` creation, code structure, etc...

**Comment by Boshen** [comment]
> Next, I'll check our current implementation. Fully or partially implemented? not yet covered?

You can assume all are partially implemented 😅 

Cross referencing the function in our code to the prettier counter part would definitely make things easier to follow.

**Comment by leaysgur** [comment]
### JS

#### AST Entry points

👀: `Format` implemented / 🚧: W.I.P. / ✨: Completed

| AST Node                                                                                                                                 |         St.          | Notes                             |
| ---------------------------------------------------------------------------------------------------------------------------------------- | :------------------: | --------------------------------- |
| RegExpLiteral                                                                                                                            |          👀          |                                   |
| BigIntLiteral                                                                                                                            |          👀          |                                   |
| NumericLiteral                                                                                                                           |          👀          |                                   |
| StringLiteral                                                                                                                            |          👀          |                                   |
| NullLiteral                                                                                                                              |          👀          |                                   |
| BooleanLiteral                                                                                                                           |          👀          |                                   |
| EmptyStatement                                                                                                                           |          👀          |                                   |
| ExpressionStatement                                                                                                                      |          👀          |                                   |
| ChainExpression                                                                                                                          |          👀          |                                   |
| ParenthesizedExpression                                                                                                                  |          👀          |                                   |
| AssignmentExpression                                                                                                                     |          👀          |                                   |
| VariableDeclarator                                                                                                                       |          👀          |                                   |
| BinaryExpression, LogicalExpression                                                                                                      |        👀, 👀        |                                   |
| AssignmentPattern                                                                                                                        |          👀          |                                   |
| MemberExpression                                                                                                                         |          👀          |                                   |
| MetaProperty                                                                                                                             |          👀          |                                   |
| Identifier                                                                                                                               |          ?           | `Identifier(Reference/Name)<'a>`? |
| SpreadElement, SpreadElementPattern, SpreadPropertyPattern, RestElement                                                                  |     👀, ?, 👀, ?     | `BindingRestElement<'a>`?         |
| FunctionDeclaration, FunctionExpression                                                                                                  |          👀          | `Function<'a>`                    |
| ArrowFunctionExpression                                                                                                                  |          👀          |                                   |
| YieldExpression                                                                                                                          |          👀          |                                   |
| AwaitExpression                                                                                                                          |          ?           |                                   |
| ExportDefaultDeclaration, ExportNamedDeclaration, ExportAllDeclaration                                                                   |      👀, 👀, 👀      |                                   |
| ImportDeclaration                                                                                                                        |          👀          |                                   |
| ImportSpecifier, ExportSpecifier, ImportNamespaceSpecifier, ExportNamespaceSpecifier, ImportDefaultSpecifier, ExportDefaultSpecifier | 👀, 👀, 👀, -, 👀, - |                                   |
| ImportAttribute                                                                                                                          |          👀          | `ImportAttributeKey<'a>`?         |
| Program, BlockStatement, StaticBlock                                                                                                     |      👀, 👀, 👀      |                                   |
| ClassBody                                                                                                                                |          👀          |                                   |
| ThrowStatement                                                                                                                           |          👀          |                                   |
| ReturnStatement                                                                                                                          |          👀          |                                   |
| NewExpression, ImportExpression, OptionalCallExpression, CallExpression                                                                  |    👀, 👀, -, 👀     |                                   |
| ObjectExpression, ObjectPattern, RecordExpression                                                                                        |      👀, 👀, -       |                                   |
| Property                                                                                                                                 |          ?           | `Property(Key/Definition)<'a>`?   |
| ObjectProperty                                                                                                                           |          👀          |                                   |
| ObjectMethod                                                                                                                             |          ?           | `MethodDefinition<'a>`?           |
| Decorator                                                                                                                                |          -           |                                   |
| ArrayExpression, ArrayPattern                                                                                                            |        👀, 👀        |                                   |
| SequenceExpression                                                                                                                       |          👀          |                                   |
| ThisExpression                                                                                                                           |          👀          |                                   |
| Super                                                                                                                                    |          👀          |                                   |
| Directive                                                                                                                                |          👀          |                                   |
| UnaryExpression                                                                                                                          |          👀          |                                   |
| UpdateExpression                                                                                                                         |          👀          |                                   |
| ConditionalExpression                                                                                                                    |          👀          |                                   |
| VariableDeclaration                                                                                                                      |          👀          |                                   |
| WithStatement                                                                                                                            |          👀          |                                   |
| IfStatement                                                                                                                              |          👀          |                                   |
| ForStatement                                                                                                                             |          👀          |                                   |
| WhileStatement                                                                                                                           |          👀          |                                   |
| ForInStatement                                                                                                                           |          👀          |                                   |
| ForOfStatement                                                                                                                           |          👀          |                                   |
| DoWhileStatement                                                                                                                         |          👀          |                                   |
| DoExpression                                                                                                                             |          -           |                                   |
| BreakStatement, ContinueStatement                                                                                                        |        👀, 👀        |                                   |
| LabeledStatement                                                                                                                         |          👀          |                                   |
| TryStatement                                                                                                                             |          👀          |                                   |
| CatchClause                                                                                                                              |          👀          |                                   |
| SwitchStatement                                                                                                                          |          👀          |                                   |
| SwitchCase                                                                                                                               |          👀          |                                   |
| DebuggerStatement                                                                                                                        |          👀          |                                   |
| ClassDeclaration, ClassExpression                                                                                                        |        👀, 👀        | `Class<'a>`                       |
| ClassMethod, ClassPrivateMethod, MethodDefinition                                                                                        |      👀, 👀, 👀      |                                   |
| ClassProperty, PropertyDefinition, ClassPrivateProperty, ClassAccessorProperty, AccessorProperty                                     |  👀, 👀, 👀, 👀, 👀  |                                   |
| TemplateElement                                                                                                                          |          👀          |                                   |
| TemplateLiteral                                                                                                                          |          👀          |                                   |
| TaggedTemplateExpression                                                                                                                 |          👀          |                                   |
| PrivateIdentifier                                                                                                                        |          👀          |                                   |
| PrivateName                                                                                                                              |          -           |                                   |
| ArgumentPlaceholder                                                                                                                      |          -           |                                   |
| ModuleExpression                                                                                                                         |          -           |                                   |

#### Printer APIs

👀: File and fn exits / 🚧: W.I.P. / ✨: Completed

|                  File | Function                          | St. | Notes                             |
| --------------------: | --------------------------------- | :-: | --------------------------------- |
|                 array | printArray                        | 👀  |                                   |
|                       | isConciselyPrintedArray           |  -  | `print_array_elements_concisely`? |
|        arrow-function | printArrowFunction                | 👀  |                                   |
|            assignment | printAssignment                   | 👀  |                                   |
|                       | printAssignmentExpression         | 👀  |                                   |
|                       | printVariableDeclarator           | 👀  |                                   |
|                       | isArrowFunctionVariableDeclarator |  -  |                                   |
|             binaryish | printBinaryishExpression          | 👀  |                                   |
|                       | shouldInlineLogicalExpression     | 👀  |                                   |
|                 block | printBlock                        | 👀  | `print_block_body`?               |
|       call-expression | printCallExpression               | 👀  | `is_commons_js_or_amd_call`?      |
|                 class | printClass                        | 👀  |                                   |
|                       | printClassBody                    | 👀  |                                   |
|                       | printClassMethod                  |  -  |                                   |
|                       | printClassProperty                | 👀  |                                   |
|            decorators | printClassMemberDecorators        |  -  |                                   |
|                       | printDecorators                   |  -  |                                   |
|  expression-statement | printExpressionStatement          |  -  |                                   |
|              function | printFunction                     | 👀  |                                   |
|                       | printMethod                       | 👀  |                                   |
|                       | printMethodValue                  |  -  |                                   |
|                       | printReturnStatement              |  -  | `print_return_or_throw_argument`? |
|                       | printReturnType                   |  -  |                                   |
|                       | printThrowStatement               |  -  | `print_return_or_throw_argument`? |
|                       | shouldPrintParamsWithoutParens    |  -  |                                   |
|   function-parameters | printFunctionParamteres           | 👀  |                                   |
|                       | shouldBreakFunctionParameters     |  -  |                                   |
|                       | shouldGroupFunctionParameters     | 👀  |                                   |
|                       | shouldHugTheOnlyFunctionParameter | 👀  |                                   |
|               literal | printLiteral                      |  -  |                                   |
|                member | printMemberExpression             |  -  |                                   |
|                       | printMemberLookup                 |  -  |                                   |
|                module | printExportDeclaration            | 👀  |                                   |
|                       | printImportDeclaration            |  -  |                                   |
|                       | printImportKind                   |  -  |                                   |
|                       | printModuleSpecifier              | 👀  |                                   |
|                object | printObject                       |  -  | `print_object_properties`?        |
|              property | printProperty                     |  -  | `property.rs`?                    |
|                       | printPropertyKey                  |  -  |                                   |
|             statement | printStatementSequence            | 👀  |                                   |
|      template-literal | printTemplateLiteral              | 👀  |                                   |
|                       | printTaggedTemplateLiteral        |  -  |                                   |
|               ternary | printTernary                      | 👀  |                                   |
|             semicolon | shouldPrintLeadingSemicolon       |  -  |                                   |
| call-arguments(utils) | printCallArguments                | 👀  |                                   |
|   member-chain(utils) | printMemberChain                  |  -  |                                   |
|                  misc | adjustClause                      | 👀  |                                   |
|                       | printAbstractToken                |  -  |                                   |
|                       | printBindExpressionCallee         |  -  |                                   |
|                       | printDeclareToken                 |  -  |                                   |
|                       | printDefiniteToken                |  -  |                                   |
|                       | printFunctionTypeParameters       |  -  |                                   |
|                       | printOptionalToken                |  -  |                                   |
|                       | printRestSpread                   |  -  |                                   |
|                       | printTypeScriptAccessibilityToken |  -  |                                   |

### JSX, TS

TBD, later... 🥲 

---

I will investigate again and update `-` and `?` things later. Then, start checking it is completed or not.

**Comment by leaysgur** [comment]
> How to handle parentheses

The basic flow is:

- When parsing the AST, `ParenthesizedExpression` is generally removed
- However, not all of them are removed
- There are cases where they should be preserved, such as JSDoc casts like `/** @type {Type} */ (x)`
  - In the first place, when using a parser compliant with ESTree, `ParenthesizedExpression` does not exist
  - Therefore, even in Prettier, the behavior changes when using parsers other than Babel...
    - https://github.com/prettier/prettier/issues/8171
- Then, when converting to IR, re-add `()` if necessary
- This is determined by `needs_parens()`.
  - https://github.com/prettier/prettier/blob/3.4.2/src/language-js/needs-parens.js

The current issues in `oxc_prettier`:

1: It assumes specifying `preserve_parens: false` as an option for `oxc_parser`.

However, as mentioned above, this cannot keep the `()` that should be preserved.

2: `needs_parens()` check is done in the `wrap!` macro, but it is not applied to all AST nodes.

Basically it can be implemented for each AST node with default: false like Biome does, but...

And for `wrap!` macro, I think collecting nodes and printing comments+parens can be decoupled. I will revisit this later.

---

For the time being, I'd like to work on updating the table above. Maybe next week, as soon as I can find the time. 💪🏻

**Comment by MichaReiser** [comment]
I'm probably a bit late here but what we did for ruff (Python formatter) was to vendor `biome_formatter`, adjust it to use our own AST (you don't want to convert between different representations for best performance). We, obviously, couldn't reuse any of the formatting logic because it's Python ;)

**Comment by leaysgur** [comment]
@MichaReiser Thanks for the very useful information!

It's not too late at all. I'm still asking myself every day whether to base on Prettier or `biome_formatter`. 😅 

So far, I mainly checked if it is possible to use `biome_formatter` directly as an external crate instead of vendor(fork)ing it, and concluded that it seems not to be possible.

On top of that,

- fork `biome_formatter` and start from scratch
- or continue to improve the existing Prettier-based code

To determine which directions is more promising, I spent time to diff the current code with Prettier, examine what is missing while picking low-hanging fruit. (And it's been two months already... 🍂  )

The situation might be different for JS(+JSX+TS) and Python, but do you think it would also be better for OXC to be based on `biome_formatter` like `ruff`? Do you have any advice?

**Comment by jycouet** [comment]
Would love to try the prototype from npm 😇

**Comment by MichaReiser** [comment]
> It's not too late at all. I'm still asking myself every day whether to base on Prettier or biome_formatter. 😅

Yeah, it's a hard decision! 

> The situation might be different for JS(+JSX+TS) and Python, but do you think it would also be better for OXC to be based on biome_formatter like ruff? Do you have any advice? 👀

I can tell you why we decided to use `biome_formatter`. Ultimately, that's a decision you have to make for your project and I'm probably biased because I wrote most of `biomejs_formatter`.

* Using `biomejs_formatter` gave us a huge head start because it gave us most infrastructure for free. The only thing we had to do was fix the compile errors (because of CST dependencies). 
* I authored most of `biomejs_formatter`. I'm obviously happy with most its design decisions and understand it well. You might not agree with some of my design decisions or have ideas to make the formatter even faster. 
* `biomejs_formatter` reimplements Prettier very closely (the semantics are 99.99% the same, it's mainly that I named things differently and that it's Rusty)

Our approach was to copy over `biomejs_formatter` and then rip out/replace everything that didn't fit well into our architecture. That worked fairly well. We ultimately also ended up introducing new IR elements that allow us to format some common Python syntax more performantly. 

You could also take a middle ground and copy code more incrementally. That could also be a good learning experience. Start with the `Printer` and the `IR` (which is, by the way, the most important foundation already). Then go ahead and change the IR however you like or make changes to the `Printer`. The advantage of this is that you already get most of the IR semantics right (it took me a long time to reverse engineer all the semantics!) It also gives you the freedom to disagree with all my naming decisions and rename the IR elements or even decide to use an entirely different representation (struct of arrays?). You could even decide that the Printer is just not what you want and that means you have your decision not to use biomejs_formatter.

From there, you can move on and vendor more of biomejs (or decide to go different ways). Vendor the `Format` trait or decide that you want to use standalone functions with the signature `fn format_node(node: &Node, context: &Context) -> FormatResult`. Rip out the `invalid-syntax` error if that's something you don't want to support... You can also decide that you aren't interested in having two separate crates (`biomejs_formatter` and `biomejs_javascript_formatter`). This removes the need for all those `FormatWith` traits because the orphan rules no longer apply (https://github.com/astral-sh/ruff/blob/0945803427ef971724ce29565cffb58ffc06166c/crates/ruff_python_formatter/orphan_rules_in_the_formatter.svg#L16)


You can then move on to start investigating comments. Biome comes with a `CommentsMap,` but maybe that's not needed for OXC or there's a better representation for OXC. 


I hope you find some of this helpful. Overall, I encourage you to take what you find useful and throw away/rewrite everything that you don't. You can also take a look at ruff's formatter to see how we changed `biomejs_formatter` to work with an AST instead of CST. Let me know if you have any questions (here or in the discord formatter channel).

**Comment by MichaReiser** [comment]
> However, while the IR and Printer could be mostly ported as-is, I'm concerned that the amount of code has turned out to be larger than expected to port IR builders.

Can you tell more about which parts your concerned? Or what has turned out to be larger than expected?

I saw a question around source text in the git repo. You may want to take a look at the `FormatElement`s in Ruff (not a CST)

https://github.com/astral-sh/ruff/blob/b63c2e126b928fab4180eda5d9132552beeee7fe/crates/ruff_formatter/src/format_element.rs#L36-L50

I saw another question around `lines_before`. The way we solve this in Ruff is by doing ad-hoc re-lexing. This is similar to what Prettier does

https://github.com/astral-sh/ruff/blob/c9fdb1f5e3a4d0d60b4537741f2c9c19e2426038/crates/ruff_python_trivia/src/tokenizer.rs#L66-L113

> FYI: This is also a WIP, but the numbers are quite interesting. 👀

This is interesting indeed and might be a good and very easy improvement for downstream users

**Comment by MichaReiser** [comment]
> It was mainly about things related to the Buffer trait and BufferExtensions, BufferSnapshot, Recording, etc.

Prettier doesn't have `Buffer` trait because each formatting function returns the formatted IR elements. This is somewhat bad for performance, which is why the formatter instead uses the `Buffer` trait where functions can write the IR elements to, which avoids allocating a `Vec` for each format function only to return the elements. It's very similar to rust's `Write` trait. 

If I were you, I'd probably delete `BufferExtension`, `BufferSnapshot` and `Recording` for now. You'll see why they exist when you start implementing specific formatting rules. But for a short overview:

* `BufferExtension`: It's mainly syntax sugar to make some convenience methods directly available on `Buffer` (which `Formatter` implements`). It follows [Rust's extension trait](https://rust-lang.github.io/rfcs/0445-extension-trait-conventions.html) approach
* `BufferSnapshot`: I don't remember which specific syntax it is used for but I think it's lambda formatting. Either way, there's one syntax where Prettier tries to format it using one layout but it may "fail" if it realize that this layout looks poorly. The way this is implemented in Prettier is by raising an exception and catching it. In biome, this is implemented by returning a `FormatError` when seeing that the layout looks poorly. However, biome has an additional challenge due to using `Buffer`: The `Buffer` now contains format elements using the "poor layout" and they need to be discarded. The way this is implemented is by first taking a snapshot of the buffer (remembering how many elements were written thus far) before trying the first layout. The buffer gets restored (reset the buffer length to the number of elements before trying the first formatting) when the `FormatError::PoorLayout` gets caught. 
* `Recording`: It's sometimes necessary to inspect the IR generated by some formatting function to make a layout decision. `Recording` allows you to do that with low overhead (without any extra allocations). The alternative is to use `Memoized` but that comes at an extra cost.

**Comment by leaysgur** [comment]
Thank you, I've learned a lot from you!

> If I were you, I'd probably delete BufferExtension, BufferSnapshot and Recording

Actually, I hadn't copied them in the initial stage.

However, since they seemed to be used in builders related to `BlockIndent` that would likely be used frequently, I ended up copying them somehow.

> https://github.com/biomejs/biome/blob/aaa9443c551dcea220adae1d26494b9850e1849e/crates/biome_formatter/src/builders.rs#L1574-L1583

More importantly, I'm glad I learned how `FormatError` is used. I now realize that I need to restore them which completly removed... 😧

**Comment by MichaReiser** [comment]
Oh, I wasn't aware that it is used in `block_indent`. It's mainly an ergonomics feature where it avoids writing the `indent` if there are no arguments. I'm not sure if I'd implement it the same again. You could consider removing it and re-add it if the explicit checks at the call sites become too annoying.

**Comment by PodaruDragos** [comment]
Hello, apologies if this has been discussed before but would the formatter be opinionated as prettier ? 
Or do we can have config files to define rules for formatting ? That can also accept feedback from the community

**Comment by leaysgur** [comment]
Close this old issue, moving forward to #10179.
